### PR TITLE
fix: handles device fingerprint iframe loading errors

### PR DESCRIPTION
**Pull request by mauriciocastillosilva-okta**
## Description:

- If iframe does not load for any reason, we now have a timeout to proceed with the flow

Resolves: OKTA-305633

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Added unit tests?
- [ ] Added e2e tests
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:
 getImage:
devicefingerprint-429-getImage

login:
devicefingerprint-429-login

### Issue:

- [OKTA-OKTA-305633](https://oktainc.atlassian.net/browse/OKTA-OKTA-305633)

### CODE REVIEW

File: src/util/DeviceFingerprint.js [hunk 1]
```diff
@@ -89,6 +92,11 @@ define(['q', 'okta'], function (Q, Okta) {
       });
       element.append($iframe);
 
+      iFrameTimeout = setTimeout(() => {
+        // If the iFrame does not load, throw an error
+        handleError('service not available');
+      }, 1000);
```
Comment (magizh-okta) [inline src/util/DeviceFingerprint.js:98]: I am worried if this is very less. Can we make this higher and disable sign in btn ?

Comment (mauriciocastillosilva-okta) [reply src/util/DeviceFingerprint.js]: I changed it to 2 sec, but I am gathering data to figure out what is the best

File: src/views/primary-auth/PrimaryAuthForm.js [hunk 2]
```diff
@@ -62,12 +62,18 @@ define([
             if (!self.settings.get('features.deviceFingerprinting')) {
               return;
             }
+            self.options.appState.trigger('loading', true);
+            self.state.trigger('togglePrimaryAuthButton', true);
```
Comment (magizh-okta) [inline src/views/primary-auth/PrimaryAuthForm.js:66]: 💯

File: test/unit/spec/DeviceFingerprint_spec.js [hunk 3]
```diff
@@ -50,6 +50,32 @@ function (Expect, $sandbox, DeviceFingerprint) {
         });
     });
 
+    it('fails if the iframe does not load', function (done) {
+      DeviceFingerprint.generateDeviceFingerprint(baseUrl, $sandbox)
+        .then(function () {
+          done.fail('Fingerprint promise should have been rejected');
+        })
+        .catch(function (reason) {
+          expect(reason).toBe('service not available');
```
Comment (magizh-okta) [inline test/unit/spec/DeviceFingerprint_spec.js:59]: we can add a check to see if setTimeout is called with 2000 ?

File: test/unit/spec/PrimaryAuth_spec.js [hunk 4]
```diff
@@ -1394,6 +1465,22 @@ function (Q, OktaAuth, LoginUtil, Okta, Util, AuthContainer, PrimaryAuthForm, Be
             expect(test.form.securityBeacon().css('background-image')).toMatch(/\/base\/target\/img\/security\/default.*\.png/);
           });
       });
+      itp('shows beacon-loading animation while loading security image (with deviceFingerprint)', function () {
+        return setup({ features: { securityImage: true, deviceFingerprinting: true }})
+          .then(function (test) {
+            test.securityBeacon = test.router.header.currentBeacon.$el;
+            spyOn(test.securityBeacon, 'toggleClass');
+            test.setNextResponse(resSecurityImage);
```
Comment (magizh-okta) [inline test/unit/spec/PrimaryAuth_spec.js:1473]: Dont we show loading when we make call to get the actual image. 

I would think that we make dfp call - toggle bacon loading state  
once dfp call is back get make getImage call - toggle bacon loading state

Comment (mauriciocastillosilva-okta) [reply test/unit/spec/PrimaryAuth_spec.js:1473]: we don't for the getImage request, we only do for device fingerprint
so this test is to know the beacon changed to loading and back. This happens during the deviceFingerprint call.

I guess we could improve this even more by adding loading animation while getImage request is made too. I'll file a follow up jira

Comment (magizh-okta) [reply test/unit/spec/PrimaryAuth_spec.js:1473]: cool might be tricky to do that since we dont do an api call it is just a get image call

**Comment by mauriciocastillosilva-okta** [comment]
Closing this PR, changes are now in https://github.com/okta/okta-signin-widget/pull/1249

**Pull request closed without merging**